runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -282,68 +331,92 @@ impl<T: Config> Pallet<T> {
 
 		let now = <frame_system::Pallet<T>>::block_number();
 
-		let mut candidate_weight = backed_candidates_weight::<T>(&backed_candidates);
-		let mut bitfields_weight = signed_bitfields_weight::<T>(signed_bitfields.len());
-		let disputes_weight = dispute_statements_weight::<T>(&disputes);
+		let candidates_weight = backed_candidates_weight::<T>(&backed_candidates);
+		let bitfields_weight = signed_bitfields_weight::<T>(signed_bitfields.len());
+		let disputes_weight = multi_dispute_statement_sets_weight::<T, _, _>(&disputes);
+
+		let current_session = <shared::Pallet<T>>::session_index();
 
 		let max_block_weight = <T as frame_system::Config>::BlockWeights::get().max_block;
 
-		METRICS.on_before_filter(candidate_weight + bitfields_weight + disputes_weight);
+		METRICS.on_before_filter(candidates_weight + bitfields_weight + disputes_weight);
+
+		T::DisputesHandler::deduplicate_and_sort_dispute_data(&mut disputes)
+			.map_err(|_e| Error::<T>::DisputeStatementsUnsortedOrDuplicates)?;
+
+		let (checked_disputes, total_consumed_weight) = {
+			// Obtain config params..
+			let config = <configuration::Pallet<T>>::config();
+			let max_spam_slots = config.dispute_max_spam_slots;
+			let post_conclusion_acceptance_period =
+				config.dispute_post_conclusion_acceptance_period;
+
+			let verify_dispute_sigs = if let FullCheck::Yes = full_check {
+				VerifyDisputeSignatures::Yes
+			} else {
+				VerifyDisputeSignatures::Skip
+			};
 
-		// Potentially trim inherent data to ensure processing will be within weight limits
-		let total_weight = {
-			if candidate_weight
+			// .. and prepare a helper closure.
+			let dispute_set_validity_check = move |set| {
+				T::DisputesHandler::filter_dispute_data(
+					set,
+					max_spam_slots,
+					post_conclusion_acceptance_period,
+					verify_dispute_sigs,
+				)
+			};
+
+			// In case of an overweight block, consume up to the entire block weight
+			// in disputes, since we will never process anything else, but invalidate
+			// the block. It's still reasonable to protect against a massive amount of disputes.
+			if candidates_weight
 				.saturating_add(bitfields_weight)
 				.saturating_add(disputes_weight) >
 				max_block_weight
 			{
-				// if the total weight is over the max block weight, first try clearing backed
-				// candidates and bitfields.
+				log::warn!("Overweight para inherent data reached the runtime {:?}", parent_hash);
+				// Technically this is not necessary, since we bail at the end of the scope
+				// still good for clarity and foot-gun prevention.
```

Comment:
is this comment outdated now?

Reply:
Well, not really. It's still a bit odd that there is something hitting the on-chain logic that is overweight. This should never happen.

Reply:
Ah, yes, you are absolutely correct.